### ISSUE TITLE
Async validation enhancements

### ISSUE THREAD

**Issue by SteveSandersonMS** [aspnet/AspNetCore#7680]
Follows on from #7614

Design ideas:

     * - Add method: editContext.AddValidationTask(FieldIdentifier f, Task t)
     *   It adds the task to a HashSet<Task> on both the FieldState and the EditContext,
     *   so we can easily get all the tasks for a given field and across the whole EditContext
     *   Also it awaits the task completion, and then regardless of outcome (success/fail/cancel),
     *   it removes the task from those hashsets.
     * - Add method: editContext.WhenAllValidationTasks()
     *   Add method: editContext.WhenAllValidationTasks(FieldIdentifier f)
     *   These return Task.WhenAll(hashSet.Values), or Task.Completed if there are none
     * - Optionally also add editContext.HasPendingValidationTasks()
     * - Add method: editContext.ValidateAsync() that awaits all the validation tasks then
     *   returns true if there are no validation messages, false otherwise
     * - Now a validation library can register tasks whenever it starts an async validation process,
     *   can cancel them if it wants, and can still issue ValidationResultsChanged notifications when
     *   each task completes. So a UI can determine whether to show "pending" state on a per-field
     *   and per-form basis, and will re-render as each field's results arrive.
     * - Note: it's unclear why we'd need WhenAllValidationTasks(FieldIdentifier) (i.e., per-field),
     *   since you wouldn't "await" this to get per-field updates (rather, you'd use ValidationResultsChanged).
     *   Maybe WhenAllValidationTasks can be private, and only called by ValidateAsync. We just expose
     *   public HasPendingValidationTasks (per-field and per-edit-context).

Will implement this shortly after getting more of the system in place, assuming it still appears to be the correct design.

**Comment by mkArtakMSFT** [comment]
@SteveSandersonMS, how important is this to be done in 3.0? Can we do this post 3.0 without breaking customers?

**Comment by SteveSandersonMS** [comment]
@mkArtakMSFT Yes. Until we make built-in support, it would be extra work for developers to track the set of pending tasks associated with form validation so they can block submission until they are all done, but it is technically possible to do it.

**Comment by IvanJosipovic** [comment]
@SteveSandersonMS, this issue is blocking using FluentValidation with Async Validators like http calls etc, which are crucial for complex validations. See this issue, https://github.com/Blazored/FluentValidation/issues/38

Do you have any guidance on how to fix this?

**Comment by isobel-cullen** [comment]
I would like the ability to perform some async action when the context is changed, at the moment I cannot use `OnFieldChanged` without locking up my application during the async call.

**Comment by AntMaster7** [comment]
@SteveSandersonMS 

How about making an async OnValidationRequested event handler in the EditContext? This way the HandleSubmitAsync method of the EditForm which is already async, could invoke and await an async form validation. This async form validation is in the end only triggered when the form is actually being submitted and not on each field changed event. Its then up to a custom validator component to notify the UI about an ongoing validation progress. I created a code example in issue #51501.

**Comment by AntMaster7** [comment]
@SteveSandersonMS Any update on this?

**Comment by austins** [comment]
Asynchronous OnValidationRequested support would be good, especially now that we have Blazor static SSR. For apps where all forms are handled in this rendering model without interactivity, it would be beneficial.

**Comment by ScarletKuro** [comment]
Any chance this could at least be considered for a Copilot candidate? This enhancement was planned about six years ago, when validation was introduced (https://github.com/dotnet/aspnetcore/pull/7614, see "Further enhancements planned for after preview 3"), but this got lost because Blazor was still in an early stage.

**Comment by AntMaster7** [comment]
> Any chance this could at least be considered for a Copilot candidate? This enhancement was planned about six years ago, when validation was introduced ([#7614](https://github.com/dotnet/aspnetcore/pull/7614)), but this got lost because Blazor was still in an early stage.

I would love to see that feature being implemented too. It got already 34 upvotes. Apparently there is a similar issue on the backend side too: https://github.com/dotnet/aspnetcore/issues/31905.

**Comment by MarvinKlein1508** [comment]
@danroth27 @javiercn I missed the Community Standup this time. I have a question regarding this feature. Will this work out of the box with `EditForm` and `DataAnnotationsValidator`? I remember when I used asnyc methods within FluentValidation and they have always caused a deadlock because `editForm` was always calling `Validate` by default instead of `ValidateAsync`

**Comment by danroth27** [comment]
> [@danroth27](https://github.com/danroth27) [@javiercn](https://github.com/javiercn) I missed the Community Standup this time. I have a question regarding this feature. Will this work out of the box with `EditForm` and `DataAnnotationsValidator`? I remember when I used asnyc methods within FluentValidation and they have always caused a deadlock because `editForm` was always calling `Validate` by default instead of `ValidateAsync`

@MarvinKlein1508 Yes, I think that's a reasonable expectation.

**Comment by danroth27** [comment]
@oroztocil We should work with @halter73 on this since he's handling the Minimal API side: https://github.com/dotnet/aspnetcore/issues/64609

**Comment by oroztocil** [comment]
@danroth27 @halter73 Good! But is there anyone working on the System.ComponentModel side of things? Since meaningful async support is blocked by the old validation infrastructure being fully synchronous. (That is the case if we talk about adding features for Minimal APIs that go above the whole Microsoft.Extensions.Validation API being fully-async already. For UI framework like Blazor we can work on asynchronously reporting the ongoing validation to the user.)
